File: frontend/src/types/memory.ts
/**
 * Memory-related TypeScript types matching backend schemas.
 */

import type { AgentVersion } from './version';

/**
 * Memory type enum - cognitive memory model.
 */
export type MemoryType = 'working' | 'short_term' | 'long_term' | 'procedural';

/**
 * Memory suggestion status.
 */
export type SuggestionStatus = 'pending' | 'approved' | 'rejected';

/**
 * Labels and descriptions for memory types.
 */
export const MEMORY_TYPE_INFO: Record<MemoryType, { label: string; description: string }> = {
  working: {
    label: 'Working Memory',
    description: 'Current session context, auto-managed by the system',
  },
  short_term: {
    label: 'Short-Term Memory',
    description: 'Persists during session, cleared when session ends',
  },
  long_term: {
    label: 'Long-Term Memory',
    description: 'Permanent storage, persists across all sessions',
  },
  procedural: {
    label: 'Procedural Memory',
    description: 'How-to knowledge and procedures in markdown format',
  },
};

/**
 * Memory types that can be manually created by users.
 */
export const CREATABLE_MEMORY_TYPES: MemoryType[] = ['short_term', 'long_term', 'procedural'];

/**
 * Request body for creating a memory.
 */
export interface MemoryCreate {
  name: string;
  content: string;
  description?: string;
  memory_type?: MemoryType;
}

/**
 * Request body for updating a memory.
 */
export interface MemoryUpdate {
  name?: string;
  content?: string;
  description?: string;
  memory_type?: MemoryType;
}

/**
 * Memory response from API.
 */
export interface Memory {
  id: string;
  name: string;
  description: string | null;
  content: string | null;
  source_path: string | null;
  memory_type: MemoryType | null;
  created_at: string | null;
}

/**
 * Response from memory create/update endpoints.
 */
export interface MemoryResponse {
  memory: Memory;
  new_version: AgentVersion;
}

/**
 * Memory suggestion from agent during conversation.
 */
export interface MemorySuggestion {
  id: string;
  agent_id: string;
  deployment_id: string | null;
  suggested_name: string;
  suggested_content: string;
  suggested_type: MemoryType;
  source_message_id: string | null;
  status: SuggestionStatus;
  created_at: string;
  reviewed_at: string | null;
  reviewed_by: string | null;
}

/**
 * Request body for reviewing a memory suggestion.
 */
export interface MemorySuggestionReview {
  status: 'approved' | 'rejected';
}

/**
 * List response for memory suggestions.
 */
export interface MemorySuggestionListResponse {
  data: MemorySuggestion[];
  total: number;
}

/**
 * Working memory state for a deployment.
 */
export interface WorkingMemoryState {
  items: WorkingMemoryItem[];
  user_injected: WorkingMemoryItem[];
}

/**
 * Single item in working memory.
 */
export interface WorkingMemoryItem {
  id: string;
  content: string;
  source: 'system' | 'user' | 'conversation';
  created_at: string;
}

/**
 * Request to inject context into working memory.
 */
export interface WorkingMemoryInject {
  content: string;
}

File: backend/app/schemas/memory.py
"""Memory management schemas for agent knowledge base."""

from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from typing import Optional
from enum import Enum

from app.schemas.version import ComponentResponse, VersionResponse


class MemoryType(str, Enum):
    """Memory component types following cognitive memory model."""
    working = "working"        # Current task context, auto-cleared
    short_term = "short_term"  # Single session, auto-expired
    long_term = "long_term"    # Persistent across sessions
    procedural = "procedural"  # How-to knowledge, markdown-based


class SuggestionStatus(str, Enum):
    """Status of a memory suggestion."""
    pending = "pending"
    approved = "approved"
    rejected = "rejected"


class MemoryCreate(BaseModel):
    """Schema for creating a new memory entry."""

    name: str = Field(..., min_length=1, max_length=255, description="Title of the memory")
    content: str = Field(..., min_length=1, description="Markdown content of the memory")
    description: Optional[str] = Field(
        None,
        max_length=500,
        description="Brief description (auto-generated from first line if not provided)",
    )
    memory_type: MemoryType = Field(
        default=MemoryType.long_term,
        description="Type of memory: working, short_term, long_term, procedural"
    )


class MemoryUpdate(BaseModel):
    """Schema for updating an existing memory entry."""

    name: Optional[str] = Field(None, min_length=1, max_length=255)
    content: Optional[str] = Field(None, min_length=1)
    description: Optional[str] = Field(None, max_length=500)
    memory_type: Optional[MemoryType] = Field(None, description="Type of memory")


class MemoryResponse(BaseModel):
    """Response schema for a single memory entry."""

    id: UUID
    name: str
    description: Optional[str]
    content: Optional[str]
    source_path: Optional[str]
    memory_type: Optional[str] = None
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class MemoryCreateResponse(BaseModel):
    """Response when creating a memory (includes new version info)."""

    memory: ComponentResponse
    new_version: VersionResponse


class MemoryUpdateResponse(BaseModel):
    """Response when updating a memory (includes new version info)."""

    memory: ComponentResponse
    new_version: VersionResponse


class MemoryDeleteResponse(BaseModel):
    """Response when deleting a memory."""

    deleted: bool
    new_version: VersionResponse


# =============================================================================
# Memory Suggestion Schemas
# =============================================================================


class MemorySuggestionCreate(BaseModel):
    """Schema for creating a memory suggestion (from agent during conversation)."""

    suggested_name: str = Field(..., min_length=1, max_length=255)
    suggested_content: str = Field(..., min_length=1)
    suggested_type: MemoryType = Field(default=MemoryType.long_term)
    source_message_id: Optional[UUID] = Field(
        None, description="ID of the message that triggered this suggestion"
    )


class MemorySuggestionResponse(BaseModel):
    """Schema for memory suggestion response."""

    id: UUID
    agent_id: UUID
    deployment_id: Optional[UUID] = None
    suggested_name: str
    suggested_content: str
    suggested_type: str
    source_message_id: Optional[UUID] = None
    status: str
    created_at: datetime
    reviewed_at: Optional[datetime] = None
    reviewed_by: Optional[UUID] = None

    class Config:
        from_attributes = True


class MemorySuggestionReview(BaseModel):
    """Schema for reviewing (approve/reject) a memory suggestion."""

    status: SuggestionStatus = Field(
        ..., description="Must be 'approved' or 'rejected'"
    )


class MemorySuggestionListResponse(BaseModel):
    """List response for memory suggestions with pagination metadata."""

    data: list[MemorySuggestionResponse]
    total: int
